Fix dependencies expansion: handle active=True correctly

joao-p-marques opened the pull request:
`active=True` in a module's manifest is a deprecated equivalent to `auto_install=True`:
https://github.com/odoo/odoo/commit/2cc7f0dd49a6be1920cef232e6c010fbafa68270#diff-367842087e68d6811cf53f2ab3eb21103cd8e793b1648ef3e57f220ee0ccde4eL345

It does not mean that Odoo installs it automatically, as mentioned in https://github.com/acsone/click-odoo-contrib/pull/9

Instead, we should handle that key the same way we handle the `auto_install` one.

### To reproduce

In v13 or v14, there is a localization module with `active` set to `True`: https://github.com/odoo/odoo/blob/13.0/addons/l10n_fi/__manifest__.py#L39
If we did `click-odoo-initdb -n devel -m base`, `l10n_fi` was added to the list of modules to generate the hash, along with its dependencies (one of them is `account`). However, after the initialization of the DB, Odoo did not install that module, which made that hash incorrect.
If we, then, removed the DB (`click-odoo-dropdb devel`) and initialized it again with `account` (`click-odoo-initdb -n devel -m base,account`) the result was that a DB cache was restored (because the previous hash included `account`) when, however, `account` was never installed in the DB.

@Tecnativa
TT24972

ping @Yajo

joao-p-marques commented:
@sbidoul are the test errors related to this change? 
I see that they fail for `master` too: https://github.com/acsone/click-odoo-contrib/runs/2064361001?check_suite_focus=true

sbidoul commented:
The test failures are not related. We'll need a new test for this case though.

joao-p-marques commented:
Ok, I added a simple test with this situation in particular, I don't know if it is the best way of testing this, but it test at least this specific situation.

sbidoul commented:
@joao-p-marques thanks for the fix! Can you rebase and add a `newfragments/93.bugfix` with a changelog entry ?

I don't know why we decided to handle active differently than auto_install in #9. That's weird.

joao-p-marques commented:
Ok @sbidoul Thanks!

I just added `newfragments/93.bugfix` to my commit. Do I need to do anything else to update the changelog?

sbidoul commented:
> Do I need to do anything else to update the changelog?

No that's fine, it will be updated when releasing.

Still I have doubts that auto_install and active mean exactly the same. For instance in the [Odoo documentation](https://www.odoo.com/documentation/14.0/reference/module.html), they are distinct entries. Is that a documentation bug ? I'd really like a test that reproduces the bug.

joao-p-marques commented:
> Still I have doubts that auto_install and active mean exactly the same. For instance in the Odoo documentation, they are distinct entries. Is that a documentation bug ? I'd really like a test that reproduces the bug.

I think the documentation is buggy, at least from reading the code I mention above.

The test I add should assert that, at least for v13 and v14.

Yajo commented:
I'm patching the docs in https://github.com/odoo/documentation/pull/1109 to clarify this situation. This fix has been in production for months, is all green and has proper docs and tests. IMHO this can be merged.

joao-p-marques commented:
@sbidoul any news on this? Could it be merged?

sbidoul commented:
I'm still have doubts. For instance if l10n_fi is active = auto_install = True, why does it not auto install when we install account ?

joao-p-marques commented:
It has been some time since I've looked into it, but I think what you are saying is exactly what I am doing here.

If `active` = `auto_install` = `True`, the module will be installed if we install `account` and should be included in the hash.

 However, the current behavior, is that it is **always** included, even if its dependencies are not installed (like `account`).

sbidoul commented:
What I mean is that l10n_fi does *not* auto install in my databases when I do "odoo -i account", so active must be doing something different than auto_install ?

joao-p-marques commented:
Oh, I see... I think it is because Odoo defaults `auto_install` to `False` [here](https://github.com/odoo/odoo/blob/13.0/odoo/modules/module.py#L335) and because of that, never actually gets [here](https://github.com/odoo/odoo/blob/13.0/odoo/modules/module.py#L370) to read the `active` property. 

I guess that is another bug, as by looking at the code you can see they should have the same behavior...

In that case, we can either leave this as is, or simply ignore the `active` keyword completely, as Odoo clearly does the same... The current behavior in `click-odoo-contrib`, however, is not correct IMHO.

Yajo commented:
It seems the docs have settled in https://github.com/odoo/documentation/pull/1427. I'd say the patch is incomplete because it doesn't contemplate situations where `active` or `auto_install` is an empty or a non-empty list of modules, where behaviours would be still different.

joao-p-marques commented:
Thanks for the reference @Yajo 

I will work on the necessary changes :+1:

joao-p-marques commented:
@Yajo @sbidoul taking into account the latest updates to the documentation, and also what I mentioned in https://github.com/acsone/click-odoo-contrib/pull/93#issuecomment-934390996 about the actual usage of the `active` keyword, I ended up:

1. Ignoring the `active` keyword completely, as Odoo does the same. For further reference, AFAICS Odoo completely removes the `active` keyword in v15: https://github.com/odoo/odoo/blob/15.0/odoo/modules/module.py
2. Applying the needed patch to handle the cases where `auto_install` is a set, although it was not possible to make a test as Odoo doesn't have a single implementation of it in the current modules.

joao-p-marques committed:
Fix dependencies expansion: handle active and auto_install keywords correctly

According to the updated documentation, active=True in a module's manifest is a deprecated equivalent to auto_install=True:
https://github.com/odoo/odoo/commit/2cc7f0dd49a6be1920cef232e6c010fbafa68270#diff-367842087e68d6811cf53f2ab3eb21103cd8e793b1648ef3e57f220ee0ccde4eL345
https://www.odoo.com/documentation/15.0/developer/reference/backend/module.html

it does not mean that Odoo installs it automatically, as mentioned in https://github.com/acsone/click-odoo-contrib/pull/9

OTOH, currently Odoo ignores it. See:
https://github.com/odoo/odoo/blob/13.0/odoo/modules/module.py#L335
https://github.com/odoo/odoo/blob/13.0/odoo/modules/module.py#L370

So, we should ignore that key.

Besides, this implements the more complex functionality of defining auto_install as a subset of the module's dependencies, as also explained in the documentation.

Also, add a new simple test for this case.

Yajo commented:
This has been more than half a year in production. Can't we merge?

Yajo commented:
Hello @sbidoul! 👋🏼 

Any chances for a new review and merge?

sbidoul commented on click_odoo_contrib/manifest.py:
```diff
@@ -69,18 +69,22 @@ def add_deps(name):
             manifest = read_manifest(module_path)
-            if manifest.get("auto_install"):
+            # "auto_install" can be a list or a boolean
+            # see Odoo's documentation for Module Manifests
```
This part seems unrelated to the issue, and should be best done in another PR.
BTW, where is this documented?

sbidoul replied:
NVM, I found the comment in the Odoo codebase.

sbidoul commented on click_odoo_contrib/manifest.py:
```diff
@@ -69,18 +69,22 @@ def add_deps(name):
+            if (
+                isinstance(auto_install, list)
+                and (
+                    manifest.get("auto_install") == []
```
This does not look right. If `auto_install` is an empty list, the module should be installed unconditionally, so in that case it should be `add_deps(module_name)`.

pedrobaeza commented:
OK, seeing that this is not direct at all, we can leave it on hold, and I'll come back to this later with less workload, and also not being something terrible. I haven't found a manifest with `active` for years...

Thank you very much for reviewing it. I'll tell you when the changes are done.

pedrobaeza commented:
I think this can be closed.
